locator/tablets.hh

```diff
@@ -376,6 +376,7 @@ public:
 
     /// Calls a given function for each tablet in the map in token ownership order.
     future<> for_each_tablet(seastar::noncopyable_function<void(tablet_id, const tablet_info&)> func) const;
+    future<> for_each_tablet_gently(seastar::noncopyable_function<future<>(tablet_id, const tablet_info&)> func) const;
```

Comment:
The existing for_each_tablet is also gentle. Both variants could be called "for_each_tablet" and rely on overload resolution.

Reply:
```
service/storage_service.cc:4434:19: error: call to member function 'for_each_tablet' is ambiguous
    co_await tmap.for_each_tablet([&] (locator::tablet_id id, const locator::tablet_info& info) -> future<> {
             ~~~~~^~~~~~~~~~~~~~~
./locator/tablets.hh:378:14: note: candidate function
    future<> for_each_tablet(seastar::noncopyable_function<void(tablet_id, const tablet_info&)> func) const;
             ^
./locator/tablets.hh:379:14: note: candidate function
    future<> for_each_tablet(seastar::noncopyable_function<future<>(tablet_id, const tablet_info&)> func) const;
```

Reply:
aha, it's ambiguous because noncopyable_function is happy to cast a function that returns non-void into a function that returns void.

Let's just change the original one to accept a future, and adjust callers to return a ready future.

Reply:
There are multiple usages this function. I'd suggest doing it as a follow up to convert and drop the original version. Sounds good?

```
$ gg for_each_tablet
locator/load_sketch.hh:77:            co_await tmap.for_each_tablet([&] (tablet_id tid, const tablet_info& ti) {
locator/tablets.cc:229:future<> tablet_map::for_each_tablet(seastar::noncopyable_function<void(tablet_id, const tablet_info&)> func) const {
locator/tablets.cc:238:future<> tablet_map::for_each_tablet_gently(seastar::noncopyable_function<future<>(tablet_id, const tablet_info&)> func) const {
locator/tablets.cc:464:        co_await tmap.for_each_tablet([this_host, &valid] (locator::tablet_id tid, const tablet_info& tinfo) {
locator/tablets.hh:378:    future<> for_each_tablet(seastar::noncopyable_function<void(tablet_id, const tablet_info&)> func) const;
locator/tablets.hh:379:    future<> for_each_tablet_gently(seastar::noncopyable_function<future<>(tablet_id, const tablet_info&)> func) const;
repair/repair.cc:2067:        co_await tmap.for_each_tablet([&] (locator::tablet_id id, const locator::tablet_info& info) {
service/storage_service.cc:4434:    co_await tmap.for_each_tablet_gently([&] (locator::tablet_id id, const locator::tablet_info& info) -> future<> {
service/tablet_allocator.cc:633:            co_await tmap.for_each_tablet([&, table = table] (tablet_id tid, const tablet_info& ti) {
service/tablet_allocator.cc:761:            co_await tmap.for_each_tablet([&, table = table] (tablet_id tid, const tablet_info& ti) {
service/topology_coordinator.cc:843:    future<> for_each_tablet_transition(std::function<void(const locator::tablet_map&,
service/topology_coordinator.cc:941:        co_await for_each_tablet_transition([&] (const locator::tablet_map& tmap,
test/boost/tablets_test.cc:1019:            tmap.for_each_tablet([&](auto tid, auto& tinfo) {
```

Reply:
Just send a patch now to convert it.

Reply:
OK. I added one in this series to convert.